packages/app/src/systems/Core/components/CoinInput.tsx

```diff
@@ -77,3 +63,43 @@ export const CoinInput = forwardRef<HTMLInputElement, CoinInputProps>(
     );
   }
 );
+
+function useDisplayedCoins(
+  initialValue: string,
+  onChange: CoinInputProps["onChange"]
+): [string, (e: React.ChangeEvent<HTMLInputElement>) => void] {
+  const [value, setValue] = useState(initialValue);
+
+  useEffect(() => {
+    if (value !== initialValue) onChange?.(value);
+  }, [value]);
+
+  useEffect(() => {
+    if (parseFloat(value) === 0 && initialValue === ZERO_AMOUNT.value) return;
+    setValue(initialValue);
+  }, [initialValue]);
+
+  const valueSetter = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      if (e.currentTarget.value.startsWith(".")) {
```

Comment:
I think we could move this code to a function, it helps to make the code cleaner.

Reply:
Please take a look at the new implementation. `useDisplayedCoins` is a helper hook that encapsulates value getting/setting and IMO the function in `valueSetter`'s `useCallback` hook is small, straightforward enough, and directly tied to `useDisplayedCoins` so much that separating it out would bring too much indirection for my taste.